flink-end-to-end-tests/flink-dataset-allround-test/src/main/java/org/apache/flink/batch/tests/Generator.java

```diff
@@ -0,0 +1,147 @@
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.io.GenericInputSplit;
+import org.apache.flink.core.io.InputSplitAssigner;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+
+/**
+ * InputFormat that generates a deterministic DataSet of Tuple2(String, Integer)
+ * <ul>
+ *     <li>String: key, can be repeated.</li>
+ *     <li>Integer: uniformly distributed int between 0 and 127</li>
+ * </ul>
+ *
+ * <p>If control path was provided, as long as this file is empty dummy elements with value equal to -1 will be emitted.
+ */
+public class Generator implements InputFormat<Tuple2<String, Integer>, GenericInputSplit> {
+
+	// total number of records
+	private final long numRecords;
+	// total number of keys
+	private final long numKeys;
+
+	// records emitted per partition
+	private long recordsPerPartition;
+	// number of keys per partition
+	private long keysPerPartition;
+
+	// number of currently emitted records
+	private long recordCnt;
+
+	// id of current partition
+	private int partitionId;
+
+	private final boolean infinite;
+
+	public Generator(long numKeys, int recordsPerKey) {
+		this(numKeys, recordsPerKey, false);
+	}
+
+	private Generator(long numKeys, int recordsPerKey, boolean infinite) {
+		this.numKeys = numKeys;
+		this.numRecords = numKeys * recordsPerKey;
+		this.infinite = infinite;
+	}
+
+	public static Generator infinite() {
+		return new Generator(Long.MAX_VALUE, 1, true);
+	}
+
+	@Override
+	public void configure(Configuration parameters) { }
+
+	@Override
+	public BaseStatistics getStatistics(BaseStatistics cachedStatistics) {
+		return null;
+	}
+
+	@Override
+	public GenericInputSplit[] createInputSplits(int minNumSplits) {
+
+		GenericInputSplit[] splits = new GenericInputSplit[minNumSplits];
+		for (int i = 0; i < minNumSplits; i++) {
+			splits[i] = new GenericInputSplit(i, minNumSplits);
+		}
+		return splits;
+	}
+
+	@Override
+	public InputSplitAssigner getInputSplitAssigner(GenericInputSplit[] inputSplits) {
+		return new DefaultInputSplitAssigner(inputSplits);
+	}
+
+	@Override
+	public void open(GenericInputSplit split) throws IOException {
+		this.partitionId = split.getSplitNumber();
+		// total number of partitions
+		int numPartitions = split.getTotalNumberOfSplits();
+
+		// ensure even distribution of records and keys
+		Preconditions.checkArgument(
+			numRecords % numPartitions == 0,
+			"Records cannot be evenly distributed among partitions");
+		Preconditions.checkArgument(
+			numKeys % numPartitions == 0,
+			"Keys cannot be evenly distributed among partitions");
+
+		this.recordsPerPartition = numRecords / numPartitions;
+		this.keysPerPartition = numKeys / numPartitions;
+
+		this.recordCnt = 0;
+	}
+
+	@Override
+	public boolean reachedEnd() {
+		return this.recordCnt >= this.recordsPerPartition;
+	}
+
+	@Override
+	public Tuple2<String, Integer> nextRecord(Tuple2<String, Integer> reuse) throws IOException {
+
+		if (infinite) {
```

Comment:
I think we should either modify `reachedEnd` to `this.recordCnt >= this.recordsPerPartition || infinite` or introduce a separate generator class, as in the current state the infinite/finite behaviors are entirely different.